<!DOCTYPE html>
<html>
<head>
  <title>GeneValidator: identify problems with gene predictions</title>
  <meta content="text/html; charset=utf-8" http-equiv="Content-Type">

  <script type="text/javascript" src="files/js/jquery-2.1.1.min.js"></script>
  <script type="text/javascript" src="files/js/bootstrap.min.js"></script>
  <script type="text/javascript" src="files/js/jquery.tablesorter.min.js"></script>
  <script type="text/javascript" src="files/js/d3.v3.min.js"></script>
  <script type="text/javascript" src="files/js/script.js"></script>
  <script type="text/javascript" src="files/js/plots.js"></script>

  <link href="files/css/bootstrap.min.css" rel="stylesheet" type="text/css">
  <link href="files/css/font-awesome.min.css" rel="stylesheet" type="text/css">
  <link href="files/css/style.css" rel="stylesheet" type="text/css">

</head>
<body>
  <div class="container">
    <span class="menu_icon pull-right"><a href="#" id="show_all_plots" onclick="show_all_plots(this);"><i class="fa fa-2x fa-bar-chart-o"></i><br>Show All Charts</a></span>
    <span class="menu_icon pull-right"><a href="https://github.com/monicadragan/GeneValidator" target="_blank"><i class="fa fa-2x fa-github"></i><br>Source Code</a></span>
    <span class="menu_icon pull-right"><a href="http://wurmlab.github.io/tools/genevalidator/documentation/v1/"><i class="fa fa-2x fa-info-circle"></i><br>About</a></span>
    <div class="clearfix"></div>
    <img class="logo" src="files/img/gene.png" alt="logo">
    <div class="page-title">
      <h1>Gene Validator <small>v1.3.7</small></h1>
      <h4 class="subheading">Identify Problems with Gene Predictions</h4>
    </div>
    <br/><br/>
    <div id="report">
      <div id="report_1"></div>
    </div>
    <br><br>
      <table id="sortable_table" class="table table-collapsed table-bordered table-condensed tablesorter">
        <thead>
          <tr id="header">
            <th>#</th>
            <th>Ranking</th>
            <th>Sequence Definition
              <span data-toggle="tooltip" title="Query definition as it apears in the input fasta file." data-placement="top"><i class="fa fa-question-circle"></i></span>
            </th>
            <th>No. Hits
              <span data-toggle="tooltip" title="Number of non-identical hits found by BLAST." data-placement="top"><i class="fa fa-question-circle"></i></span>
            </th>
                        <th class="sorter-false">
              <b>Length Cluster</b>
                                  <span data-toggle="tooltip" title="Charts available for this validation" data-placement="top"><i class="fa fa-bar-chart-o chartIcon"></i></span>&nbsp;<span data-toggle="tooltip" title="Check whether the prediction length fits most of the BLAST hit lengths, by 1D hierarchical clusterization. Meaning of the output displayed: Query_length [Main Cluster Length Interval]" data-placement="top"><i class="fa fa-question-circle"></i></span>
                            </th>
                        <th class="sorter-false">
              <b>Length Rank</b>
                                   <span data-toggle="tooltip" title="Check whether the rank of the prediction length lies among 80% of all the BLAST hit lengths." data-placement="top"><i class="fa fa-question-circle"></i></span>
                            </th>
                        <th class="sorter-false">
              <b>Gene Merge</b>
                                  <span data-toggle="tooltip" title="Charts available for this validation" data-placement="top"><i class="fa fa-bar-chart-o chartIcon"></i></span>&nbsp;<span data-toggle="tooltip" title="Check whether BLAST hits make evidence about a merge of two genes that match the predicted gene." data-placement="top"><i class="fa fa-question-circle"></i></span>
                            </th>
                        <th class="sorter-false">
              <b>Duplication</b>
                                   <span data-toggle="tooltip" title="Check whether there is a duplicated subsequence in the predicted gene by counting the hsp residue coverage of the prediction, for each hit." data-placement="top"><i class="fa fa-question-circle"></i></span>
                            </th>
                        <th class="sorter-false">
              <b></b>
                                   <span data-toggle="tooltip" title="" data-placement="top"><i class="fa fa-question-circle"></i></span>
                            </th>
                        <th class="sorter-false">
              <b></b>
                                   <span data-toggle="tooltip" title="" data-placement="top"><i class="fa fa-question-circle"></i></span>
                            </th>
                        <th class="sorter-false">
              <b>Missing/Extra sequences</b>
                                  <span data-toggle="tooltip" title="Charts available for this validation" data-placement="top"><i class="fa fa-bar-chart-o chartIcon"></i></span>&nbsp;<span data-toggle="tooltip" title="Finds missing and extra sequences in the prediction, based on the multiple alignment of the best hits. Also counts the percentage of the conserved regions that appear in the prediction." data-placement="top"><i class="fa fa-question-circle"></i></span>
                            </th>
                        <th class="sorter-false chart-column"></th>
          </tr>
        </thead>
        <tbody><tr>
  <td>1</td>
  <td data-score="0"><div class="ratings"><div class="empty-stars"></div><div class="full-stars" style="width:0.0%;"></div></div></td>
  <td title="Definition" class="seq_definition">GB10034-PA  (amel_OGSv1.0_pep.fa_34)</td>
  <td title="No. Hits">3</td>
      <td title="Length Cluster" class="warning">Not enough evidence</td>
        <td title="Length Rank" class="warning">Not enough evidence</td>
        <td title="Gene Merge" class="warning">Not enough evidence</td>
        <td title="Duplication" class="warning">Not enough evidence</td>
        <td title="" class="warning"></td>
        <td title="" class="warning"></td>
        <td title="Missing/Extra sequences" class="warning">Not enough evidence</td>
  
<td></td>
</tr>
<tr class="tablesorter-childRow" style="display:none;">
  <td colspan="14">
    <div></div>
  </td>
</tr>
<tr>
  <td>2</td>
  <td data-score="38"><div class="ratings"><div class="empty-stars"></div><div class="full-stars" style="width:32.3%;"></div></div></td>
  <td title="Definition" class="seq_definition">GB10040-PA  (amel_OGSv1.0_pep.fa_40)</td>
  <td title="No. Hits">23</td>
      <td title="Length Cluster" class="danger my-btn-danger" data-approach="If the query sequence is well conserved and similar sequences (BLAST hits) are correct, we can expect query and hit sequences to have similar lengths. Here, we cluster the lengths of hit sequences and compare the length of our query sequence to the most dense cluster of hit lengths. " data-explanation="The most dense length-cluster of BLAST hits includes sequences that are from 694 to 810 amino-acids long. The query sequence is 1070 amino-acids long and is thus outside the most dense length-cluster of BLAST hits." data-conclusion="This suggests that the query sequence may be too long." onclick="showDiv(this, 'toggle2');
          addPlot('toggle2', 'files/json/all_validations_prot.fasta_2_len_clusters.json', 'bars', 'Length Cluster Validation: Distribution of BLAST hit lengths', 'Query Sequence, black;Most Dense Cluster,red;Other Hits, blue', 'Sequence Length', 'Number of Sequences', '1070', '');
    AddExplanation(this, 'toggle2');">1070&nbsp;[694,&nbsp;810]</td>
        <td title="Length Rank" class="success" data-approach="If the query sequence is well conserved and similar sequences (BLAST hits) are correct, we can expect query and hit sequences to have similar lengths." data-explanation="The query sequence is  1070 amino-acids long. BLAST identified 23 hit sequences with lengths from 694 to 1285 amino-acids (median: 805; mean: 877). 5 of these hit sequences (i.e. 22%) are shorter than the query sequence." data-conclusion="There is no reason to believe there is any problem with the length of the query sequence." onclick="showDiv(this, 'toggle2'); AddExplanation(this,'toggle2');">22%</td>
        <td title="Gene Merge" class="success my-btn-success" data-approach="We expect the query sequence to encode a single protein-coding gene. Here, we analyse the High-scoring Segment Pairs (HSPs) identified by BLAST to determine whether the query includes sequence from two or more genes." data-explanation="The start coordinates and the end coordinates of HSPs are unimodally distributed." data-conclusion="This suggest that the query sequence represents a single gene." onclick="showDiv(this, 'toggle2');
          addPlot('toggle2', 'files/json/all_validations_prot.fasta_2_match_2d.json', 'scatter', 'Gene Merge Validation: Start/end of matching hit coord. on query (1 point/hit)', '', 'Start Offset (most left hsp)', 'End Offset (most right hsp)', '', '');
          addPlot('toggle2', 'files/json/all_validations_prot.fasta_2_match.json', 'lines', 'Gene Merge Validation: Query coord covered by blast hit (1 line/hit)', '', 'Offset in Prediction', 'Hit Number', '23', '');
    AddExplanation(this, 'toggle2');">0.0</td>
        <td title="Duplication" class="success" data-approach="We expect each BLAST hit to match each region of the query at most once. Here, we calculate the distribution of hit coverage against the query sequence and use the Wilcoxon test to determine if it is higher than 1." data-explanation="The Wilcoxon test produced a p-value of 1." data-conclusion="This suggests that the query sequence contains no erroneous duplications." onclick="showDiv(this, 'toggle2'); AddExplanation(this,'toggle2');">1.0</td>
        <td title="" class="warning"></td>
        <td title="" class="warning"></td>
        <td title="Missing/Extra sequences" class="danger my-btn-danger" data-approach="We expect the query sequence to be similar to the top ten BLAST hits. Here, we create a statistical consensus model of those top hits and compare the query to this model." data-explanation="The query sequence includes 93.82716049382715% amino-acid residues present in the consensus model. 16.2890385793019% of residues in the query sequence are absent from the consensus profile.  21.922841396203307% of residues in the consensus profile are absent from the query sequence." data-conclusion="These results suggest that there may be some problems with the query sequence. The query sequence has a high percentage (21.922841396203307%) of missing residues when compared to the statistical profile (the cut-off is 20%)." onclick="showDiv(this, 'toggle2');
          addPlot('toggle2', 'files/json/all_validations_prot.fasta_2_ma.json', 'align', 'Missing/Extra sequences Validation: Multiple Align. & Statistical model of hits', 'Conserved Region, Yellow', 'Offset in the Alignment', '', '12', 'Prediction, hit&nbsp;1, hit&nbsp;2, hit&nbsp;3, hit&nbsp;4, hit&nbsp;5, hit&nbsp;6, hit&nbsp;7, hit&nbsp;8, hit&nbsp;9, hit&nbsp;10, Statistical Model');
    AddExplanation(this, 'toggle2');">93.82716049382715%&nbsp;conserved; 16.2890385793019%&nbsp;extra; 21.922841396203307%&nbsp;missing.</td>
  
  <td>
    <button title="Show plots" name="plot_btn" class="plot_btn btn btn-default" onclick="showDiv(this, 'toggle2');
                   addPlot('toggle2', 'files/json/all_validations_prot.fasta_2_len_clusters.json', 'bars', 'Length Cluster Validation: Distribution of BLAST hit lengths', 'Query Sequence, black;Most Dense Cluster,red;Other Hits, blue', 'Sequence Length', 'Number of Sequences', '1070', '');                                   addPlot('toggle2', 'files/json/all_validations_prot.fasta_2_match_2d.json', 'scatter', 'Gene Merge Validation: Start/end of matching hit coord. on query (1 point/hit)', '', 'Start Offset (most left hsp)', 'End Offset (most right hsp)', '', ''); addPlot('toggle2', 'files/json/all_validations_prot.fasta_2_match.json', 'lines', 'Gene Merge Validation: Query coord covered by blast hit (1 line/hit)', '', 'Offset in Prediction', 'Hit Number', '23', '');                                                       addPlot('toggle2', 'files/json/all_validations_prot.fasta_2_ma.json', 'align', 'Missing/Extra sequences Validation: Multiple Align. & Statistical model of hits', 'Conserved Region, Yellow', 'Offset in the Alignment', '', '12', 'Prediction, hit&nbsp;1, hit&nbsp;2, hit&nbsp;3, hit&nbsp;4, hit&nbsp;5, hit&nbsp;6, hit&nbsp;7, hit&nbsp;8, hit&nbsp;9, hit&nbsp;10, Statistical Model');          "><i class="fa fa-bar-chart-o"></i></button>
  </td>
</tr>
<tr class="tablesorter-childRow" name="plot_row" style="display:none;">
  <td colspan="14" id="toggle2row">
    <div id="toggle2" class="expanded-child"></div>
  </td>
</tr>
<tr>
  <td>3</td>
  <td data-score="67"><div class="ratings"><div class="empty-stars"></div><div class="full-stars" style="width:56.949999999999996%;"></div></div></td>
  <td title="Definition" class="seq_definition">GB10056-PA </td>
  <td title="No. Hits">10</td>
      <td title="Length Cluster" class="success my-btn-success" data-approach="If the query sequence is well conserved and similar sequences (BLAST hits) are correct, we can expect query and hit sequences to have similar lengths. Here, we cluster the lengths of hit sequences and compare the length of our query sequence to the most dense cluster of hit lengths. " data-explanation="The most dense length-cluster of BLAST hits includes sequences that are from 317 to 365 amino-acids long. The query sequence is 339 amino-acids long and is thus inside the most dense length-cluster of BLAST hits." data-conclusion="There is no reason to believe there is any problem with the length of the query sequence." onclick="showDiv(this, 'toggle3');
          addPlot('toggle3', 'files/json/all_validations_prot.fasta_3_len_clusters.json', 'bars', 'Length Cluster Validation: Distribution of BLAST hit lengths', 'Query Sequence, black;Most Dense Cluster,red;Other Hits, blue', 'Sequence Length', 'Number of Sequences', '339', '');
    AddExplanation(this, 'toggle3');">339&nbsp;[317,&nbsp;365]</td>
        <td title="Length Rank" class="success" data-approach="If the query sequence is well conserved and similar sequences (BLAST hits) are correct, we can expect query and hit sequences to have similar lengths." data-explanation="The query sequence is  339 amino-acids long. BLAST identified 10 hit sequences with lengths from 166 to 365 amino-acids (median: 331; mean: 313). 4 of these hit sequences (i.e. 40%) are shorter than the query sequence." data-conclusion="There is no reason to believe there is any problem with the length of the query sequence." onclick="showDiv(this, 'toggle3'); AddExplanation(this,'toggle3');">40%</td>
        <td title="Gene Merge" class="success my-btn-success" data-approach="We expect the query sequence to encode a single protein-coding gene. Here, we analyse the High-scoring Segment Pairs (HSPs) identified by BLAST to determine whether the query includes sequence from two or more genes." data-explanation="The distribution of start and/or end-coordinates of HSPs are multi-modal. To detect potential problems we performed a linear regression (with coordinates weighted inversely proportionally to  hit strength). The resulting slope is -3.5." data-conclusion="This slope is  outside our empirically calculated thresholds (0.4 and 1.2). There is no evidence that the query contains sequence from multiple genes." onclick="showDiv(this, 'toggle3');
          addPlot('toggle3', 'files/json/all_validations_prot.fasta_3_match_2d.json', 'scatter', 'Gene Merge Validation: Start/end of matching hit coord. on query (1 point/hit)', '', 'Start Offset (most left hsp)', 'End Offset (most right hsp)', '352.3021657808258', '-3.500833679692545');
          addPlot('toggle3', 'files/json/all_validations_prot.fasta_3_match.json', 'lines', 'Gene Merge Validation: Query coord covered by blast hit (1 line/hit)', '', 'Offset in Prediction', 'Hit Number', '10', '');
    AddExplanation(this, 'toggle3');">-3.5</td>
        <td title="Duplication" class="success" data-approach="We expect each BLAST hit to match each region of the query at most once. Here, we calculate the distribution of hit coverage against the query sequence and use the Wilcoxon test to determine if it is higher than 1." data-explanation="The Wilcoxon test produced a p-value of 1." data-conclusion="This suggests that the query sequence contains no erroneous duplications." onclick="showDiv(this, 'toggle3'); AddExplanation(this,'toggle3');">1.0</td>
        <td title="" class="warning"></td>
        <td title="" class="warning"></td>
        <td title="Missing/Extra sequences" class="warning">Mafft error</td>
  
  <td>
    <button title="Show plots" name="plot_btn" class="plot_btn btn btn-default" onclick="showDiv(this, 'toggle3');
                   addPlot('toggle3', 'files/json/all_validations_prot.fasta_3_len_clusters.json', 'bars', 'Length Cluster Validation: Distribution of BLAST hit lengths', 'Query Sequence, black;Most Dense Cluster,red;Other Hits, blue', 'Sequence Length', 'Number of Sequences', '339', '');                                   addPlot('toggle3', 'files/json/all_validations_prot.fasta_3_match_2d.json', 'scatter', 'Gene Merge Validation: Start/end of matching hit coord. on query (1 point/hit)', '', 'Start Offset (most left hsp)', 'End Offset (most right hsp)', '352.3021657808258', '-3.500833679692545'); addPlot('toggle3', 'files/json/all_validations_prot.fasta_3_match.json', 'lines', 'Gene Merge Validation: Query coord covered by blast hit (1 line/hit)', '', 'Offset in Prediction', 'Hit Number', '10', '');                                                  "><i class="fa fa-bar-chart-o"></i></button>
  </td>
</tr>
<tr class="tablesorter-childRow" name="plot_row" style="display:none;">
  <td colspan="14" id="toggle3row">
    <div id="toggle3" class="expanded-child"></div>
  </td>
</tr>
<tr>
  <td>4</td>
  <td data-score="50"><div class="ratings"><div class="empty-stars"></div><div class="full-stars" style="width:42.5%;"></div></div></td>
  <td title="Definition" class="seq_definition">GB10058-PA </td>
  <td title="No. Hits">114</td>
      <td title="Length Cluster" class="danger my-btn-danger" data-approach="If the query sequence is well conserved and similar sequences (BLAST hits) are correct, we can expect query and hit sequences to have similar lengths. Here, we cluster the lengths of hit sequences and compare the length of our query sequence to the most dense cluster of hit lengths. " data-explanation="The most dense length-cluster of BLAST hits includes sequences that are from 1106 to 1567 amino-acids long. The query sequence is 1052 amino-acids long and is thus outside the most dense length-cluster of BLAST hits." data-conclusion="This suggests that the query sequence may be too short." onclick="showDiv(this, 'toggle4');
          addPlot('toggle4', 'files/json/all_validations_prot.fasta_4_len_clusters.json', 'bars', 'Length Cluster Validation: Distribution of BLAST hit lengths', 'Query Sequence, black;Most Dense Cluster,red;Other Hits, blue', 'Sequence Length', 'Number of Sequences', '1052', '');
    AddExplanation(this, 'toggle4');">1052&nbsp;[1106,&nbsp;1567]</td>
        <td title="Length Rank" class="success" data-approach="If the query sequence is well conserved and similar sequences (BLAST hits) are correct, we can expect query and hit sequences to have similar lengths." data-explanation="The query sequence is  1052 amino-acids long. BLAST identified 114 hit sequences with lengths from 455 to 1787 amino-acids (median: 1168; mean: 1108). 46 of these hit sequences (i.e. 40%) are longer than the query sequence." data-conclusion="There is no reason to believe there is any problem with the length of the query sequence." onclick="showDiv(this, 'toggle4'); AddExplanation(this,'toggle4');">40%</td>
        <td title="Gene Merge" class="success my-btn-success" data-approach="We expect the query sequence to encode a single protein-coding gene. Here, we analyse the High-scoring Segment Pairs (HSPs) identified by BLAST to determine whether the query includes sequence from two or more genes." data-explanation="The distribution of start and/or end-coordinates of HSPs are multi-modal. To detect potential problems we performed a linear regression (with coordinates weighted inversely proportionally to  hit strength). The resulting slope is -0.25." data-conclusion="This slope is  outside our empirically calculated thresholds (0.4 and 1.2). There is no evidence that the query contains sequence from multiple genes." onclick="showDiv(this, 'toggle4');
          addPlot('toggle4', 'files/json/all_validations_prot.fasta_4_match_2d.json', 'scatter', 'Gene Merge Validation: Start/end of matching hit coord. on query (1 point/hit)', '', 'Start Offset (most left hsp)', 'End Offset (most right hsp)', '1053.589103259333', '-0.25314129112975275');
          addPlot('toggle4', 'files/json/all_validations_prot.fasta_4_match.json', 'lines', 'Gene Merge Validation: Query coord covered by blast hit (1 line/hit)', '', 'Offset in Prediction', 'Hit Number', '114', '');
    AddExplanation(this, 'toggle4');">-0.25</td>
        <td title="Duplication" class="success" data-approach="We expect each BLAST hit to match each region of the query at most once. Here, we calculate the distribution of hit coverage against the query sequence and use the Wilcoxon test to determine if it is higher than 1." data-explanation="The Wilcoxon test produced a p-value of 1." data-conclusion="This suggests that the query sequence contains no erroneous duplications." onclick="showDiv(this, 'toggle4'); AddExplanation(this,'toggle4');">1.0</td>
        <td title="" class="warning"></td>
        <td title="" class="warning"></td>
        <td title="Missing/Extra sequences" class="warning">Mafft error</td>
  
  <td>
    <button title="Show plots" name="plot_btn" class="plot_btn btn btn-default" onclick="showDiv(this, 'toggle4');
                   addPlot('toggle4', 'files/json/all_validations_prot.fasta_4_len_clusters.json', 'bars', 'Length Cluster Validation: Distribution of BLAST hit lengths', 'Query Sequence, black;Most Dense Cluster,red;Other Hits, blue', 'Sequence Length', 'Number of Sequences', '1052', '');                                   addPlot('toggle4', 'files/json/all_validations_prot.fasta_4_match_2d.json', 'scatter', 'Gene Merge Validation: Start/end of matching hit coord. on query (1 point/hit)', '', 'Start Offset (most left hsp)', 'End Offset (most right hsp)', '1053.589103259333', '-0.25314129112975275'); addPlot('toggle4', 'files/json/all_validations_prot.fasta_4_match.json', 'lines', 'Gene Merge Validation: Query coord covered by blast hit (1 line/hit)', '', 'Offset in Prediction', 'Hit Number', '114', '');                                                  "><i class="fa fa-bar-chart-o"></i></button>
  </td>
</tr>
<tr class="tablesorter-childRow" name="plot_row" style="display:none;">
  <td colspan="14" id="toggle4row">
    <div id="toggle4" class="expanded-child"></div>
  </td>
</tr>
<tr>
  <td>5</td>
  <td data-score="100"><div class="ratings"><div class="empty-stars"></div><div class="full-stars" style="width:85.0%;"></div></div></td>
  <td title="Definition" class="seq_definition">GB10063-PA </td>
  <td title="No. Hits">200</td>
      <td title="Length Cluster" class="success my-btn-success" data-approach="If the query sequence is well conserved and similar sequences (BLAST hits) are correct, we can expect query and hit sequences to have similar lengths. Here, we cluster the lengths of hit sequences and compare the length of our query sequence to the most dense cluster of hit lengths. " data-explanation="The most dense length-cluster of BLAST hits includes sequences that are from 105 to 629 amino-acids long. The query sequence is 615 amino-acids long and is thus inside the most dense length-cluster of BLAST hits." data-conclusion="There is no reason to believe there is any problem with the length of the query sequence." onclick="showDiv(this, 'toggle5');
          addPlot('toggle5', 'files/json/all_validations_prot.fasta_5_len_clusters.json', 'bars', 'Length Cluster Validation: Distribution of BLAST hit lengths', 'Query Sequence, black;Most Dense Cluster,red;Other Hits, blue', 'Sequence Length', 'Number of Sequences', '615', '');
    AddExplanation(this, 'toggle5');">615&nbsp;[105,&nbsp;629]</td>
        <td title="Length Rank" class="success" data-approach="If the query sequence is well conserved and similar sequences (BLAST hits) are correct, we can expect query and hit sequences to have similar lengths." data-explanation="The query sequence is  615 amino-acids long. BLAST identified 200 hit sequences with lengths from 105 to 2485 amino-acids (median: 748; mean: 880). 90 of these hit sequences (i.e. 45%) are longer than the query sequence." data-conclusion="There is no reason to believe there is any problem with the length of the query sequence." onclick="showDiv(this, 'toggle5'); AddExplanation(this,'toggle5');">45%</td>
        <td title="Gene Merge" class="success my-btn-success" data-approach="We expect the query sequence to encode a single protein-coding gene. Here, we analyse the High-scoring Segment Pairs (HSPs) identified by BLAST to determine whether the query includes sequence from two or more genes." data-explanation="The distribution of start and/or end-coordinates of HSPs are multi-modal. To detect potential problems we performed a linear regression (with coordinates weighted inversely proportionally to  hit strength). The resulting slope is -0.02." data-conclusion="This slope is  outside our empirically calculated thresholds (0.4 and 1.2). There is no evidence that the query contains sequence from multiple genes." onclick="showDiv(this, 'toggle5');
          addPlot('toggle5', 'files/json/all_validations_prot.fasta_5_match_2d.json', 'scatter', 'Gene Merge Validation: Start/end of matching hit coord. on query (1 point/hit)', '', 'Start Offset (most left hsp)', 'End Offset (most right hsp)', '552.571307882806', '-0.01847710358414582');
          addPlot('toggle5', 'files/json/all_validations_prot.fasta_5_match.json', 'lines', 'Gene Merge Validation: Query coord covered by blast hit (1 line/hit)', '', 'Offset in Prediction', 'Hit Number', '200', '');
    AddExplanation(this, 'toggle5');">-0.02</td>
        <td title="Duplication" class="warning">Unexpected error</td>
        <td title="" class="warning"></td>
        <td title="" class="warning"></td>
        <td title="Missing/Extra sequences" class="warning">Mafft error</td>
  
  <td>
    <button title="Show plots" name="plot_btn" class="plot_btn btn btn-default" onclick="showDiv(this, 'toggle5');
                   addPlot('toggle5', 'files/json/all_validations_prot.fasta_5_len_clusters.json', 'bars', 'Length Cluster Validation: Distribution of BLAST hit lengths', 'Query Sequence, black;Most Dense Cluster,red;Other Hits, blue', 'Sequence Length', 'Number of Sequences', '615', '');                                   addPlot('toggle5', 'files/json/all_validations_prot.fasta_5_match_2d.json', 'scatter', 'Gene Merge Validation: Start/end of matching hit coord. on query (1 point/hit)', '', 'Start Offset (most left hsp)', 'End Offset (most right hsp)', '552.571307882806', '-0.01847710358414582'); addPlot('toggle5', 'files/json/all_validations_prot.fasta_5_match.json', 'lines', 'Gene Merge Validation: Query coord covered by blast hit (1 line/hit)', '', 'Offset in Prediction', 'Hit Number', '200', '');                                                  "><i class="fa fa-bar-chart-o"></i></button>
  </td>
</tr>
<tr class="tablesorter-childRow" name="plot_row" style="display:none;">
  <td colspan="14" id="toggle5row">
    <div id="toggle5" class="expanded-child"></div>
  </td>
</tr>
<tr>
  <td>6</td>
  <td data-score="50"><div class="ratings"><div class="empty-stars"></div><div class="full-stars" style="width:42.5%;"></div></div></td>
  <td title="Definition" class="seq_definition">GB10071-PA </td>
  <td title="No. Hits">27</td>
      <td title="Length Cluster" class="danger my-btn-danger" data-approach="If the query sequence is well conserved and similar sequences (BLAST hits) are correct, we can expect query and hit sequences to have similar lengths. Here, we cluster the lengths of hit sequences and compare the length of our query sequence to the most dense cluster of hit lengths. " data-explanation="The most dense length-cluster of BLAST hits includes sequences that are from 379 to 402 amino-acids long. The query sequence is 494 amino-acids long and is thus outside the most dense length-cluster of BLAST hits." data-conclusion="This suggests that the query sequence may be too long." onclick="showDiv(this, 'toggle6');
          addPlot('toggle6', 'files/json/all_validations_prot.fasta_6_len_clusters.json', 'bars', 'Length Cluster Validation: Distribution of BLAST hit lengths', 'Query Sequence, black;Most Dense Cluster,red;Other Hits, blue', 'Sequence Length', 'Number of Sequences', '494', '');
    AddExplanation(this, 'toggle6');">494&nbsp;[379,&nbsp;402]</td>
        <td title="Length Rank" class="danger" data-approach="If the query sequence is well conserved and similar sequences (BLAST hits) are correct, we can expect query and hit sequences to have similar lengths." data-explanation="The query sequence is  494 amino-acids long. BLAST identified 27 hit sequences with lengths from 335 to 416 amino-acids (median: 398; mean: 397). All hit sequences are shorter than the query sequence." data-conclusion="The sequence may be too long." onclick="showDiv(this, 'toggle6'); AddExplanation(this,'toggle6');">0%&nbsp;(too&nbsp;long)</td>
        <td title="Gene Merge" class="success my-btn-success" data-approach="We expect the query sequence to encode a single protein-coding gene. Here, we analyse the High-scoring Segment Pairs (HSPs) identified by BLAST to determine whether the query includes sequence from two or more genes." data-explanation="The distribution of start and/or end-coordinates of HSPs are multi-modal. To detect potential problems we performed a linear regression (with coordinates weighted inversely proportionally to  hit strength). The resulting slope is -6.98." data-conclusion="This slope is  outside our empirically calculated thresholds (0.4 and 1.2). There is no evidence that the query contains sequence from multiple genes." onclick="showDiv(this, 'toggle6');
          addPlot('toggle6', 'files/json/all_validations_prot.fasta_6_match_2d.json', 'scatter', 'Gene Merge Validation: Start/end of matching hit coord. on query (1 point/hit)', '', 'Start Offset (most left hsp)', 'End Offset (most right hsp)', '494.21460450242125', '-6.977334922898902');
          addPlot('toggle6', 'files/json/all_validations_prot.fasta_6_match.json', 'lines', 'Gene Merge Validation: Query coord covered by blast hit (1 line/hit)', '', 'Offset in Prediction', 'Hit Number', '27', '');
    AddExplanation(this, 'toggle6');">-6.98</td>
        <td title="Duplication" class="danger" data-approach="We expect each BLAST hit to match each region of the query at most once. Here, we calculate the distribution of hit coverage against the query sequence and use the Wilcoxon test to determine if it is higher than 1." data-explanation="The Wilcoxon test produced a p-value of 0.002 (average = 1.07)." data-conclusion="The null hypothesis is rejected - thus a region of the query sequence is likely repeated more than once." onclick="showDiv(this, 'toggle6'); AddExplanation(this,'toggle6');">0.0</td>
        <td title="" class="warning"></td>
        <td title="" class="warning"></td>
        <td title="Missing/Extra sequences" class="danger my-btn-danger" data-approach="We expect the query sequence to be similar to the top ten BLAST hits. Here, we create a statistical consensus model of those top hits and compare the query to this model." data-explanation="The query sequence includes 79.45205479452055% amino-acid residues present in the consensus model. 21.022727272727273% of residues in the query sequence are absent from the consensus profile.  6.0606060606060606% of residues in the consensus profile are absent from the query sequence." data-conclusion="These results suggest that there may be some problems with the query sequence. The query sequence has a high percentage (21.022727272727273%) of extra residues absent from the statistical profile (the cut-off is 20%)." onclick="showDiv(this, 'toggle6');
          addPlot('toggle6', 'files/json/all_validations_prot.fasta_6_ma.json', 'align', 'Missing/Extra sequences Validation: Multiple Align. & Statistical model of hits', 'Conserved Region, Yellow', 'Offset in the Alignment', '', '12', 'Prediction, hit&nbsp;1, hit&nbsp;2, hit&nbsp;3, hit&nbsp;4, hit&nbsp;5, hit&nbsp;6, hit&nbsp;7, hit&nbsp;8, hit&nbsp;9, hit&nbsp;10, Statistical Model');
    AddExplanation(this, 'toggle6');">79.45205479452055%&nbsp;conserved; 21.022727272727273%&nbsp;extra; 6.0606060606060606%&nbsp;missing.</td>
  
  <td>
    <button title="Show plots" name="plot_btn" class="plot_btn btn btn-default" onclick="showDiv(this, 'toggle6');
                   addPlot('toggle6', 'files/json/all_validations_prot.fasta_6_len_clusters.json', 'bars', 'Length Cluster Validation: Distribution of BLAST hit lengths', 'Query Sequence, black;Most Dense Cluster,red;Other Hits, blue', 'Sequence Length', 'Number of Sequences', '494', '');                                   addPlot('toggle6', 'files/json/all_validations_prot.fasta_6_match_2d.json', 'scatter', 'Gene Merge Validation: Start/end of matching hit coord. on query (1 point/hit)', '', 'Start Offset (most left hsp)', 'End Offset (most right hsp)', '494.21460450242125', '-6.977334922898902'); addPlot('toggle6', 'files/json/all_validations_prot.fasta_6_match.json', 'lines', 'Gene Merge Validation: Query coord covered by blast hit (1 line/hit)', '', 'Offset in Prediction', 'Hit Number', '27', '');                                                       addPlot('toggle6', 'files/json/all_validations_prot.fasta_6_ma.json', 'align', 'Missing/Extra sequences Validation: Multiple Align. & Statistical model of hits', 'Conserved Region, Yellow', 'Offset in the Alignment', '', '12', 'Prediction, hit&nbsp;1, hit&nbsp;2, hit&nbsp;3, hit&nbsp;4, hit&nbsp;5, hit&nbsp;6, hit&nbsp;7, hit&nbsp;8, hit&nbsp;9, hit&nbsp;10, Statistical Model');          "><i class="fa fa-bar-chart-o"></i></button>
  </td>
</tr>
<tr class="tablesorter-childRow" name="plot_row" style="display:none;">
  <td colspan="14" id="toggle6row">
    <div id="toggle6" class="expanded-child"></div>
  </td>
</tr>
<tr>
  <td>7</td>
  <td data-score="0"><div class="ratings"><div class="empty-stars"></div><div class="full-stars" style="width:0.0%;"></div></div></td>
  <td title="Definition" class="seq_definition">GB10091-PA </td>
  <td title="No. Hits">0</td>
      <td title="Length Cluster" class="warning">Not enough evidence</td>
        <td title="Length Rank" class="warning">Not enough evidence</td>
        <td title="Gene Merge" class="warning">Not enough evidence</td>
        <td title="Duplication" class="warning">Not enough evidence</td>
        <td title="" class="warning"></td>
        <td title="" class="warning"></td>
        <td title="Missing/Extra sequences" class="warning">Not enough evidence</td>
  
<td></td>
</tr>
<tr class="tablesorter-childRow" style="display:none;">
  <td colspan="14">
    <div></div>
  </td>
</tr>
<tr>
  <td>8</td>
  <td data-score="67"><div class="ratings"><div class="empty-stars"></div><div class="full-stars" style="width:56.949999999999996%;"></div></div></td>
  <td title="Definition" class="seq_definition">GB10098-PA </td>
  <td title="No. Hits">93</td>
      <td title="Length Cluster" class="success my-btn-success" data-approach="If the query sequence is well conserved and similar sequences (BLAST hits) are correct, we can expect query and hit sequences to have similar lengths. Here, we cluster the lengths of hit sequences and compare the length of our query sequence to the most dense cluster of hit lengths. " data-explanation="The most dense length-cluster of BLAST hits includes sequences that are from 633 to 965 amino-acids long. The query sequence is 826 amino-acids long and is thus inside the most dense length-cluster of BLAST hits." data-conclusion="There is no reason to believe there is any problem with the length of the query sequence." onclick="showDiv(this, 'toggle8');
          addPlot('toggle8', 'files/json/all_validations_prot.fasta_8_len_clusters.json', 'bars', 'Length Cluster Validation: Distribution of BLAST hit lengths', 'Query Sequence, black;Most Dense Cluster,red;Other Hits, blue', 'Sequence Length', 'Number of Sequences', '826', '');
    AddExplanation(this, 'toggle8');">826&nbsp;[633,&nbsp;965]</td>
        <td title="Length Rank" class="success" data-approach="If the query sequence is well conserved and similar sequences (BLAST hits) are correct, we can expect query and hit sequences to have similar lengths." data-explanation="The query sequence is  826 amino-acids long. BLAST identified 93 hit sequences with lengths from 250 to 1406 amino-acids (median: 798; mean: 795). 37 of these hit sequences (i.e. 40%) are shorter than the query sequence." data-conclusion="There is no reason to believe there is any problem with the length of the query sequence." onclick="showDiv(this, 'toggle8'); AddExplanation(this,'toggle8');">40%</td>
        <td title="Gene Merge" class="success my-btn-success" data-approach="We expect the query sequence to encode a single protein-coding gene. Here, we analyse the High-scoring Segment Pairs (HSPs) identified by BLAST to determine whether the query includes sequence from two or more genes." data-explanation="The distribution of start and/or end-coordinates of HSPs are multi-modal. To detect potential problems we performed a linear regression (with coordinates weighted inversely proportionally to  hit strength). The resulting slope is -0.04." data-conclusion="This slope is  outside our empirically calculated thresholds (0.4 and 1.2). There is no evidence that the query contains sequence from multiple genes." onclick="showDiv(this, 'toggle8');
          addPlot('toggle8', 'files/json/all_validations_prot.fasta_8_match_2d.json', 'scatter', 'Gene Merge Validation: Start/end of matching hit coord. on query (1 point/hit)', '', 'Start Offset (most left hsp)', 'End Offset (most right hsp)', '809.121555586734', '-0.036663940433547745');
          addPlot('toggle8', 'files/json/all_validations_prot.fasta_8_match.json', 'lines', 'Gene Merge Validation: Query coord covered by blast hit (1 line/hit)', '', 'Offset in Prediction', 'Hit Number', '93', '');
    AddExplanation(this, 'toggle8');">-0.04</td>
        <td title="Duplication" class="success" data-approach="We expect each BLAST hit to match each region of the query at most once. Here, we calculate the distribution of hit coverage against the query sequence and use the Wilcoxon test to determine if it is higher than 1." data-explanation="The Wilcoxon test produced a p-value of 1." data-conclusion="This suggests that the query sequence contains no erroneous duplications." onclick="showDiv(this, 'toggle8'); AddExplanation(this,'toggle8');">1.0</td>
        <td title="" class="warning"></td>
        <td title="" class="warning"></td>
        <td title="Missing/Extra sequences" class="warning">Mafft error</td>
  
  <td>
    <button title="Show plots" name="plot_btn" class="plot_btn btn btn-default" onclick="showDiv(this, 'toggle8');
                   addPlot('toggle8', 'files/json/all_validations_prot.fasta_8_len_clusters.json', 'bars', 'Length Cluster Validation: Distribution of BLAST hit lengths', 'Query Sequence, black;Most Dense Cluster,red;Other Hits, blue', 'Sequence Length', 'Number of Sequences', '826', '');                                   addPlot('toggle8', 'files/json/all_validations_prot.fasta_8_match_2d.json', 'scatter', 'Gene Merge Validation: Start/end of matching hit coord. on query (1 point/hit)', '', 'Start Offset (most left hsp)', 'End Offset (most right hsp)', '809.121555586734', '-0.036663940433547745'); addPlot('toggle8', 'files/json/all_validations_prot.fasta_8_match.json', 'lines', 'Gene Merge Validation: Query coord covered by blast hit (1 line/hit)', '', 'Offset in Prediction', 'Hit Number', '93', '');                                                  "><i class="fa fa-bar-chart-o"></i></button>
  </td>
</tr>
<tr class="tablesorter-childRow" name="plot_row" style="display:none;">
  <td colspan="14" id="toggle8row">
    <div id="toggle8" class="expanded-child"></div>
  </td>
</tr>
<tr>
  <td>9</td>
  <td data-score="50"><div class="ratings"><div class="empty-stars"></div><div class="full-stars" style="width:42.5%;"></div></div></td>
  <td title="Definition" class="seq_definition">GB10113-PA </td>
  <td title="No. Hits">18</td>
      <td title="Length Cluster" class="danger my-btn-danger" data-approach="If the query sequence is well conserved and similar sequences (BLAST hits) are correct, we can expect query and hit sequences to have similar lengths. Here, we cluster the lengths of hit sequences and compare the length of our query sequence to the most dense cluster of hit lengths. " data-explanation="The most dense length-cluster of BLAST hits includes sequences that are from 300 to 506 amino-acids long. The query sequence is 659 amino-acids long and is thus outside the most dense length-cluster of BLAST hits." data-conclusion="This suggests that the query sequence may be too long." onclick="showDiv(this, 'toggle9');
          addPlot('toggle9', 'files/json/all_validations_prot.fasta_9_len_clusters.json', 'bars', 'Length Cluster Validation: Distribution of BLAST hit lengths', 'Query Sequence, black;Most Dense Cluster,red;Other Hits, blue', 'Sequence Length', 'Number of Sequences', '659', '');
    AddExplanation(this, 'toggle9');">659&nbsp;[300,&nbsp;506]</td>
        <td title="Length Rank" class="success" data-approach="If the query sequence is well conserved and similar sequences (BLAST hits) are correct, we can expect query and hit sequences to have similar lengths." data-explanation="The query sequence is  659 amino-acids long. BLAST identified 18 hit sequences with lengths from 300 to 2224 amino-acids (median: 503; mean: 712). 7 of these hit sequences (i.e. 39%) are shorter than the query sequence." data-conclusion="There is no reason to believe there is any problem with the length of the query sequence." onclick="showDiv(this, 'toggle9'); AddExplanation(this,'toggle9');">39%</td>
        <td title="Gene Merge" class="success my-btn-success" data-approach="We expect the query sequence to encode a single protein-coding gene. Here, we analyse the High-scoring Segment Pairs (HSPs) identified by BLAST to determine whether the query includes sequence from two or more genes." data-explanation="The distribution of start and/or end-coordinates of HSPs are multi-modal. To detect potential problems we performed a linear regression (with coordinates weighted inversely proportionally to  hit strength). The resulting slope is -2.32." data-conclusion="This slope is  outside our empirically calculated thresholds (0.4 and 1.2). There is no evidence that the query contains sequence from multiple genes." onclick="showDiv(this, 'toggle9');
          addPlot('toggle9', 'files/json/all_validations_prot.fasta_9_match_2d.json', 'scatter', 'Gene Merge Validation: Start/end of matching hit coord. on query (1 point/hit)', '', 'Start Offset (most left hsp)', 'End Offset (most right hsp)', '372.8025039114409', '-2.323577963298711');
          addPlot('toggle9', 'files/json/all_validations_prot.fasta_9_match.json', 'lines', 'Gene Merge Validation: Query coord covered by blast hit (1 line/hit)', '', 'Offset in Prediction', 'Hit Number', '18', '');
    AddExplanation(this, 'toggle9');">-2.32</td>
        <td title="Duplication" class="success" data-approach="We expect each BLAST hit to match each region of the query at most once. Here, we calculate the distribution of hit coverage against the query sequence and use the Wilcoxon test to determine if it is higher than 1." data-explanation="The Wilcoxon test produced a p-value of 0.250." data-conclusion="This suggests that the query sequence contains no erroneous duplications." onclick="showDiv(this, 'toggle9'); AddExplanation(this,'toggle9');">0.25</td>
        <td title="" class="warning"></td>
        <td title="" class="warning"></td>
        <td title="Missing/Extra sequences" class="warning">Mafft error</td>
  
  <td>
    <button title="Show plots" name="plot_btn" class="plot_btn btn btn-default" onclick="showDiv(this, 'toggle9');
                   addPlot('toggle9', 'files/json/all_validations_prot.fasta_9_len_clusters.json', 'bars', 'Length Cluster Validation: Distribution of BLAST hit lengths', 'Query Sequence, black;Most Dense Cluster,red;Other Hits, blue', 'Sequence Length', 'Number of Sequences', '659', '');                                   addPlot('toggle9', 'files/json/all_validations_prot.fasta_9_match_2d.json', 'scatter', 'Gene Merge Validation: Start/end of matching hit coord. on query (1 point/hit)', '', 'Start Offset (most left hsp)', 'End Offset (most right hsp)', '372.8025039114409', '-2.323577963298711'); addPlot('toggle9', 'files/json/all_validations_prot.fasta_9_match.json', 'lines', 'Gene Merge Validation: Query coord covered by blast hit (1 line/hit)', '', 'Offset in Prediction', 'Hit Number', '18', '');                                                  "><i class="fa fa-bar-chart-o"></i></button>
  </td>
</tr>
<tr class="tablesorter-childRow" name="plot_row" style="display:none;">
  <td colspan="14" id="toggle9row">
    <div id="toggle9" class="expanded-child"></div>
  </td>
</tr>
<tr>
  <td>10</td>
  <td data-score="0"><div class="ratings"><div class="empty-stars"></div><div class="full-stars" style="width:0.0%;"></div></div></td>
  <td title="Definition" class="seq_definition">sp|P83011|13KDA_SCYCA  13.2 kDa protein (Fragment) OS=Scyliorhinus canicula PE=1 SV=1</td>
  <td title="No. Hits">0</td>
      <td title="Length Cluster" class="warning">Not enough evidence</td>
        <td title="Length Rank" class="warning">Not enough evidence</td>
        <td title="Gene Merge" class="warning">Not enough evidence</td>
        <td title="Duplication" class="warning">Not enough evidence</td>
        <td title="" class="warning"></td>
        <td title="" class="warning"></td>
        <td title="Missing/Extra sequences" class="warning">Not enough evidence</td>
  
<td></td>
</tr>
<tr class="tablesorter-childRow" style="display:none;">
  <td colspan="14">
    <div></div>
  </td>
</tr>
